gaps/gap-17_offline_requestor_model/gap-17_offline_requestor_model.md

```diff
@@ -0,0 +1,112 @@
+- Requestor daemon may reconnect at any time, and exercise control over the Activity via ExeScript
+
+**Notes:**
+- This scenario is only possible with payment schemes which assume upfront payment or allow long intervals between payments or with self-sustained payment
+ 
+#### **B. Requestor offline ("fire&forget")** 
+Characteristics:
+- Requestor daemon remains online until Agreement is signed and Activity started
+- Requestor daemon disconnects permanently, while the Activity continues (probably until agreed computation is complete, or funds run out)
+
+**Notes:**
+- This scenario is only possible with payment schemes which assume upfront payment or allow long intervals between payments or with self-sustained payment
+ 
+#### **C. Requestor delegates control**
+Characteristics:
+- Requestor daemon remains online until Agreement is signed and Activity started
+- Requestor daemon transfers grants control rights to another node/identity, which remains online and exercises control over the Activity 
+
+**Notes:**
+- This requires **Agreement Permissions Management** feature
+- (TBD) Not sure how payments are controlled by a “delegate”...? Are all Payment actions allowed for Payments associated with a given Agreement?
+
+### Proposed features
+
+This GAP introduces following features which are aimed at enabling the "offline Reuqestor" scenarios listed above:
+- Activity Attach/Detach
+- Self-sustained Payments
+- Agreement Permissions Management
+
+The mapping between the scenarios and features required to implement them is indicated below:
+
+| | Activity attach/detach | Self-sustained payments | Agreement Permissions Management |
+|-|-|-|-|
+| Requestor partially connected     | X | X | |
+| Requestor offline ("fire&forget") | X | X | |
+| Requestor delegates control       | X | | X |
+
+#### **Feature: Activity attach/detach**
+We propose to introduce an ability for Requestor node to disconnect from the network while controlled Activity remains active, and then gracefully reconnect, and take control of the Activity. **Note** this can happen intentionally, or as a result of eg. a network failure or software error. Therefore the ability to attach a Requestor to resume control over an Activity is improving the reliability and robustness of Golem as a platform. 
+
+The implementation of this feature requires considerations on two levels:
+
+1. Requestor Agent application goes offline, while corresponding `yagna` daemon remains online 
+    - The HL API implementations shall include the ability to obtain a "live" instance of Activity object and corresponding Agreement object based on `activityID`, and then perform actions on them (ExeScript command execution and results processing, Agreement control, including termination), as in the following pseudo-code example:
+       ```
+       ...
+       activity = Golem.attach_activity(activityId)   # returns an "attached" Activity object, which can then be used to manipulate the Activity
```

Comment:
Actually I believe this should be an explicit operation on a `Golem` object, which is representing the "context" in which the Activity resides.

Alternative 'style' would be to create an `Activity` object and populate its ID, then 'attach' it to the `Golem` context, for the engine to 'hydrate' it (reconcile against current state and populate resptective Activity properties):

```
...
activity = Activity(activityId)
activity = Golem.attach_activity(activity)
...
```

...but the 'shorthand' syntax I've proposed seems 'lighter', so I prefer that.

Reply:
Just FYI, my current version is
```
golem = GolemNode()
activity = golem.activity(activity_id)
```

Reply:
I'm not sure if I understand the question.

`GolemNode` objects corresponds to a (`YAGNA_URL`, `APP_KEY`) pair. If this is enough to control the activity, it will allow the subsequent control. It doesn't matter if the activity was created with this `GolemNode` instance or not (or in this script run, or this `yagna` instance).

IOW, I think that if you created an activity and shared with me `activity_id` and your `APP_KEY` I could do:
```
async with GolemNode() as golem:
    activity = golem.activity(activity_id)
    activity.execute_command(...)
```
without additional effort (except for setting `yagna` identity with the same `APP_KEY`).
(I might be missing something, but in principle this is how this should work, I think?).

Reply:
...OK, cool, and for example if we do this:
```
async with GolemNode() as golem:
    activity = golem.activity(activity_id)
```

...will the `activity` object correctly represent the actual state of the Activity?

Reply:
Yes, but only after one more call:
```
async with GolemNode() as golem:
    activity = golem.activity(activity_id)
    await activity.get_data()
```

Reply:
After some more consideration, I decided I don't understand what is "attach" and "detach" in your model :/ For me, every entity with `activity_id` and credentials is "attached" in the same way (i.e. it can interact with the Activity).
So - sorry if my answers don't make sense :)